lib/stateTransition/AbstractStateTransition.js

```diff
@@ -49,6 +77,154 @@ class AbstractStateTransition {
   hash() {
     return hash(this.serialize()).toString('hex');
   }
+
+  /**
+   *  Returns signature
+   *
+   * @return {string|null}
+   */
+  getSignature() {
+    return this.signature;
+  }
+
+  /**
+   * Returns public key id
+   *
+   * @returns {number|null}
+   */
+  getSignaturePublicKeyId() {
+    return this.signaturePublicKeyId;
+  }
+
+  /**
+   * Sign data with privateKey
+   *
+   * @param {IdentityPublicKey} identityPublicKey
+   * @param {string} privateKey
+   * @return {AbstractStateTransition}
+   */
+  sign(identityPublicKey, privateKey) {
+    const data = this.serialize({ skipSignature: true });
+    const privateKeyInHex = Buffer
+      .from(privateKey, 'base64')
+      .toString('hex');
+
+    switch (identityPublicKey.getType()) {
+      case IdentityPublicKey.TYPES.ECDSA_SECP256K1:
+        this.signature = signData(data, privateKeyInHex).toString('base64');
+        break;
+      default:
+        throw new InvalidSignatureTypeError(identityPublicKey.getType());
+    }
+
+    const extractedPublicKey = new PrivateKey(privateKeyInHex)
```

Comment:
We should validate it before we sign :)

Reply:
You mean inside IdentityPublicKey.TYPES.ECDSA_SECP256K, right?

Reply:
Yep